File: backend/app/api/routes/tokenizer.py
"""API routes for tokenizer operations."""

from fastapi import APIRouter, HTTPException, status, Query
from typing import Optional, List, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum

from nanotorch.tokenizer import (
    BaseTokenizer,
    CharTokenizer,
    WordTokenizer,
    BPETokenizer,
    get_tokenizer,
)

router = APIRouter(prefix="/api/v1/tokenizer", tags=["tokenizer"])


# ============================================================================
# Request/Response Models
# ============================================================================

class TokenizerType(str, Enum):
    """Supported tokenizer types."""
    char = "char"
    word = "word"
    bpe = "bpe"


class TokenizeRequest(BaseModel):
    """Request to tokenize text."""
    text: str = Field(..., description="Text to tokenize", min_length=1)
    tokenizer_type: TokenizerType = Field(
        default=TokenizerType.char,
        description="Type of tokenizer to use"
    )
    vocab_size: int = Field(
        default=10000,
        ge=100,
        le=100000,
        description="Maximum vocabulary size"
    )
    num_merges: int = Field(
        default=1000,
        ge=10,
        le=10000,
        description="Number of BPE merge operations (only for BPE tokenizer)"
    )
    training_texts: Optional[List[str]] = Field(
        default=None,
        description="Optional training texts. If not provided, uses the input text for training."
    )


class TokenPosition(BaseModel):
    """Position information for a token."""
    start_position: int = Field(..., description="Start position in original text")
    end_position: int = Field(..., description="End position in original text")


class TokenInfo(BaseModel):
    """Information about a single token."""
    token_id: int = Field(..., description="Token ID")
    text: str = Field(..., description="Token text")
    frequency: int = Field(..., description="Token frequency in training corpus")
    is_special: bool = Field(..., description="Whether this is a special token")
    start_position: Optional[int] = Field(None, description="Start position in original text")
    end_position: Optional[int] = Field(None, description="End position in original text")


class VocabularyTokenInfo(BaseModel):
    """Information about a token in the vocabulary."""
    id: int = Field(..., description="Token ID")
    text: str = Field(..., description="Token text")
    frequency: int = Field(..., description="Token frequency")
    is_special: bool = Field(..., description="Whether this is a special token")


class TokenizeResponse(BaseModel):
    """Response from tokenization."""
    success: bool = Field(..., description="Whether tokenization was successful")
    token_ids: List[int] = Field(..., description="List of token IDs")
    tokens: List[str] = Field(..., description="List of token strings")
    token_details: List[TokenInfo] = Field(..., description="Detailed token information")
    vocabulary_summary: Dict[str, Any] = Field(..., description="Vocabulary summary")
    tokenizer_type: str = Field(..., description="Type of tokenizer used")
    error: Optional[str] = Field(None, description="Error message if failed")


class VocabularyRequest(BaseModel):
    """Request to get vocabulary."""
    tokenizer_type: TokenizerType = Field(
        default=TokenizerType.char,
        description="Type of tokenizer"
    )
    vocab_size: int = Field(
        default=10000,
        ge=100,
        le=100000,
        description="Maximum vocabulary size"
    )
    training_texts: Optional[List[str]] = Field(
        default=None,
        description="Training texts. Uses default corpus if not provided."
    )


class VocabularyResponse(BaseModel):
    """Response with vocabulary data."""
    success: bool = Field(..., description="Whether request was successful")
    vocab_size: int = Field(..., description="Total vocabulary size")
    tokens: List[VocabularyTokenInfo] = Field(..., description="All tokens in vocabulary")
    special_tokens: Dict[str, int] = Field(..., description="Special token IDs")
    tokenizer_type: str = Field(..., description="Type of tokenizer")
    error: Optional[str] = Field(None, description="Error message if failed")


class TokenDetailRequest(BaseModel):
    """Request to get detailed token information."""
    token_id: int = Field(..., ge=0, description="Token ID to look up")
    tokenizer_type: TokenizerType = Field(
        default=TokenizerType.char,
        description="Type of tokenizer"
    )
    vocab_size: int = Field(
        default=10000,
        ge=100,
        le=100000,
        description="Maximum vocabulary size"
    )
    training_texts: Optional[List[str]] = Field(
        default=None,
        description="Training texts. Uses default corpus if not provided."
    )


class TokenDetailResponse(BaseModel):
    """Response with detailed token information."""
    success: bool = Field(..., description="Whether request was successful")
    token_info: Optional[Dict[str, Any]] = Field(None, description="Token information")
    tokenizer_type: str = Field(..., description="Type of tokenizer")
    error: Optional[str] = Field(None, description="Error message if failed")


class DecodeRequest(BaseModel):
    """Request to decode token IDs."""
    token_ids: List[int] = Field(..., description="Token IDs to decode")
    tokenizer_type: TokenizerType = Field(
        default=TokenizerType.char,
        description="Type of tokenizer"
    )
    vocab_size: int = Field(
        default=10000,
        ge=100,
        le=100000,
        description="Maximum vocabulary size"
    )
    training_texts: Optional[List[str]] = Field(
        default=None,
        description="Training texts. Uses default corpus if not provided."
    )


class DecodeResponse(BaseModel):
    """Response from decoding."""
    success: bool = Field(..., description="Whether decoding was successful")
    text: str = Field(..., description="Decoded text")
    tokenizer_type: str = Field(..., description="Type of tokenizer used")
    error: Optional[str] = Field(None, description="Error message if failed")


class CompareRequest(BaseModel):
    """Request to compare different tokenizers."""
    text: str = Field(..., description="Text to tokenize", min_length=1)
    vocab_size: int = Field(
        default=10000,
        ge=100,
        le=100000,
        description="Maximum vocabulary size for all tokenizers"
    )
    num_merges: int = Field(
        default=1000,
        ge=10,
        le=10000,
        description="Number of BPE merge operations"
    )
    training_texts: Optional[List[str]] = Field(
        default=None,
        description="Optional training texts"
    )


class ComparisonResult(BaseModel):
    """Result from comparing tokenizers."""
    tokenizer_type: str = Field(..., description="Tokenizer type")
    num_tokens: int = Field(..., description="Number of tokens produced")
    tokens: List[str] = Field(..., description="Token strings")
    token_ids: List[int] = Field(..., description="Token IDs")
    oov_count: int = Field(..., description="Number of out-of-vocabulary tokens")


class CompareResponse(BaseModel):
    """Response from tokenizer comparison."""
    success: bool = Field(..., description="Whether comparison was successful")
    comparisons: List[ComparisonResult] = Field(..., description="Comparison results for each tokenizer")
    error: Optional[str] = Field(None, description="Error message if failed")


# ============================================================================
# Default Training Corpus
# ============================================================================

DEFAULT_TRAINING_CORPUS = [
    "The quick brown fox jumps over the lazy dog.",
    "Machine learning is a subset of artificial intelligence.",
    "Natural language processing deals with the interaction between computers and human language.",
    "Transformers are neural network architectures that use attention mechanisms.",
    "Tokenization is the process of breaking text into smaller units called tokens.",
    "Deep learning models require large amounts of training data.",
    "Attention mechanisms allow models to focus on different parts of the input.",
    "Embeddings represent words as dense vectors in a continuous space.",
    "The Transformer architecture revolutionized natural language processing.",
    "Byte Pair Encoding is a subword tokenization algorithm.",
]


# ============================================================================
# Helper Functions
# ============================================================================

def get_or_train_tokenizer(
    tokenizer_type: str,
    training_texts: Optional[List[str]],
    vocab_size: int,
    num_merges: int = 1000
) -> BaseTokenizer:
    """
    Get an existing trained tokenizer or train a new one.

    Args:
        tokenizer_type: Type of tokenizer
        training_texts: Texts to train on (uses default if None)
        vocab_size: Maximum vocabulary size
        num_merges: Number of BPE merges

    Returns:
        Trained tokenizer instance
    """
    # Get tokenizer
    if tokenizer_type == "bpe":
        tokenizer = BPETokenizer(vocab_size=vocab_size, num_merges=num_merges)
    elif tokenizer_type == "word":
        tokenizer = WordTokenizer(vocab_size=vocab_size)
    else:
        tokenizer = CharTokenizer(vocab_size=vocab_size)

    # Train the tokenizer
    texts = training_texts if training_texts else DEFAULT_TRAINING_CORPUS
    tokenizer.train(texts)

    return tokenizer


# ============================================================================
# API Endpoints
# ============================================================================

@router.post("/tokenize", response_model=TokenizeResponse)
async def tokenize(request: TokenizeRequest):
    """Tokenize text using the specified tokenizer.

    Args:
        request: Tokenization request with text, tokenizer type, and options

    Returns:
        Tokenization response with token IDs, token strings, and metadata
    """
    try:
        # Create and train tokenizer
        tokenizer = get_or_train_tokenizer(
            tokenizer_type=request.tokenizer_type.value,
            training_texts=request.training_texts,
            vocab_size=request.vocab_size,
            num_merges=request.num_merges
        )

        # Encode text
        token_ids = tokenizer.encode(request.text)
        tokens = [tokenizer.reverse_vocab.get(tid, '<unk>') for tid in token_ids]

        # Get detailed token information with positions
        token_details_data = tokenizer.encode_with_positions(request.text)

        token_details = [
            TokenInfo(
                token_id=t['token_id'],
                text=t['text'],
                frequency=tokenizer.token_frequencies.get(t['token_id'], 0),
                is_special=t['is_special'],
                start_position=t.get('start_position'),
                end_position=t.get('end_position')
            )
            for t in token_details_data
        ]

        # Get vocabulary summary
        vocabulary = tokenizer.get_vocabulary()
        vocabulary_summary = {
            "vocab_size": len(tokenizer.reverse_vocab),
            "special_tokens": tokenizer.get_special_tokens(),
            "most_common": sorted(
                [
                    {"id": tid, "text": info["text"], "frequency": info["frequency"]}
                    for tid, info in vocabulary.items()
                    if not info["is_special"]
                ],
                key=lambda x: -x["frequency"]
            )[:10]
        }

        return TokenizeResponse(
            success=True,
            token_ids=token_ids,
            tokens=tokens,
            token_details=token_details,
            vocabulary_summary=vocabulary_summary,
            tokenizer_type=request.tokenizer_type.value
        )

    except Exception as e:
        return TokenizeResponse(
            success=False,
            token_ids=[],
            tokens=[],
            token_details=[],
            vocabulary_summary={},
            tokenizer_type=request.tokenizer_type.value,
            error=str(e)
        )


@router.post("/vocabulary", response_model=VocabularyResponse)
async def get_vocabulary(request: VocabularyRequest):
    """Get the vocabulary for a tokenizer.

    Args:
        request: Vocabulary request with tokenizer type and options

    Returns:
        Vocabulary response with all tokens and their information
    """
    try:
        # Create and train tokenizer
        tokenizer = get_or_train_tokenizer(
            tokenizer_type=request.tokenizer_type.value,
            training_texts=request.training_texts,
            vocab_size=request.vocab_size
        )

        # Get vocabulary
        vocabulary = tokenizer.get_vocabulary()

        # Convert to response format
        tokens = [
            VocabularyTokenInfo(
                id=token_id,
                text=info["text"],
                frequency=info["frequency"],
                is_special=info["is_special"]
            )
            for token_id, info in vocabulary.items()
        ]

        return VocabularyResponse(
            success=True,
            vocab_size=len(tokenizer.reverse_vocab),
            tokens=tokens,
            special_tokens=tokenizer.get_special_tokens(),
            tokenizer_type=request.tokenizer_type.value
        )

    except Exception as e:
        return VocabularyResponse(
            success=False,
            vocab_size=0,
            tokens=[],
            special_tokens={},
            tokenizer_type=request.tokenizer_type.value,
            error=str(e)
        )


@router.post("/token-detail", response_model=TokenDetailResponse)
async def get_token_detail(request: TokenDetailRequest):
    """Get detailed information about a specific token.

    Args:
        request: Token detail request

    Returns:
        Token detail response
    """
    try:
        # Create and train tokenizer
        tokenizer = get_or_train_tokenizer(
            tokenizer_type=request.tokenizer_type.value,
            training_texts=request.training_texts,
            vocab_size=request.vocab_size
        )

        # Get token info
        token_info = tokenizer.get_token_info(request.token_id)

        if token_info is None:
            return TokenDetailResponse(
                success=False,
                token_info=None,
                tokenizer_type=request.tokenizer_type.value,
                error=f"Token ID {request.token_id} not found in vocabulary"
            )

        return TokenDetailResponse(
            success=True,
            token_info=token_info,
            tokenizer_type=request.tokenizer_type.value
        )

    except Exception as e:
        return TokenDetailResponse(
            success=False,
            token_info=None,
            tokenizer_type=request.tokenizer_type.value,
            error=str(e)
        )


@router.post("/decode", response_model=DecodeResponse)
async def decode(request: DecodeRequest):
    """Decode token IDs back into text.

    Args:
        request: Decode request with token IDs and tokenizer type

    Returns:
        Decoded text
    """
    try:
        # Create and train tokenizer
        tokenizer = get_or_train_tokenizer(
            tokenizer_type=request.tokenizer_type.value,
            training_texts=request.training_texts,
            vocab_size=request.vocab_size
        )

        # Decode
        text = tokenizer.decode(request.token_ids)

        return DecodeResponse(
            success=True,
            text=text,
            tokenizer_type=request.tokenizer_type.value
        )

    except Exception as e:
        return DecodeResponse(
            success=False,
            text="",
            tokenizer_type=request.tokenizer_type.value,
            error=str(e)
        )


@router.post("/compare", response_model=CompareResponse)
async def compare_tokenizers(request: CompareRequest):
    """Compare different tokenizers on the same text.

    Args:
        request: Comparison request with text and options

    Returns:
        Comparison results for each tokenizer type
    """
    try:
        comparisons = []

        for tokenizer_type in ["char", "word", "bpe"]:
            # Create and train tokenizer
            tokenizer = get_or_train_tokenizer(
                tokenizer_type=tokenizer_type,
                training_texts=request.training_texts,
                vocab_size=request.vocab_size,
                num_merges=request.num_merges
            )

            # Encode
            token_ids = tokenizer.encode(request.text)
            tokens = [tokenizer.reverse_vocab.get(tid, '<unk>') for tid in token_ids]

            # Count OOV (unknown) tokens
            oov_count = sum(1 for tid in token_ids if tid == tokenizer.unk_token_id())

            comparisons.append(
                ComparisonResult(
                    tokenizer_type=tokenizer_type,
                    num_tokens=len(token_ids),
                    tokens=tokens,
                    token_ids=token_ids,
                    oov_count=oov_count
                )
            )

        return CompareResponse(
            success=True,
            comparisons=comparisons
        )

    except Exception as e:
        return CompareResponse(
            success=False,
            comparisons=[],
            error=str(e)
        )


@router.get("/types")
async def get_tokenizer_types():
    """Get available tokenizer types.

    Returns:
        List of available tokenizer types with descriptions
    """
    return {
        "success": True,
        "tokenizer_types": [
            {
                "type": "char",
                "name": "Character-level",
                "description": "Splits text into individual characters. Simple but results in long sequences."
            },
            {
                "type": "word",
                "name": "Word-level",
                "description": "Splits text into words using regex. Handles punctuation separately."
            },
            {
                "type": "bpe",
                "name": "Byte Pair Encoding",
                "description": "Subword tokenization that learns merge operations. Balances vocabulary size and sequence length."
            }
        ]
    }

File: frontend/src/types/tokenizer.ts
/**
 * TypeScript types for tokenizer functionality.
 *
 * This module defines all types related to tokenization, vocabulary management,
 * and tokenizer configuration.
 */

/**
 * Supported tokenizer types.
 * - char: Character-level tokenization (splits text into individual characters)
 * - word: Word-level tokenization (splits text into words using regex)
 * - bpe: Byte Pair Encoding (subword tokenization with learned merge operations)
 */
export type TokenType = 'char' | 'word' | 'bpe';

/**
 * Display information for tokenizer types.
 */
export interface TokenizerTypeInfo {
  type: TokenType;
  name: string;
  description: string;
}

/**
 * Position information for a token within the original text.
 * Used to highlight which part of the text corresponds to which token.
 */
export interface TokenPosition {
  startPosition: number;
  endPosition: number;
}

/**
 * Detailed information about a single token.
 */
export interface TokenData {
  /** Unique identifier for this token in the vocabulary */
  id: number;
  /** The text string that this token represents */
  text: string;
  /** Frequency of this token in the training corpus */
  frequency: number;
  /** Whether this is a special token (e.g., <pad>, <unk>, <sos>, <eos>) */
  isSpecial: boolean;
  /** Starting position in the original text (if applicable) */
  startPosition?: number | null;
  /** Ending position in the original text (if applicable) */
  endPosition?: number | null;
  /** Embedding vector for this token (if available) */
  embedding?: number[];
}

/**
 * Token information returned from the tokenize endpoint.
 */
export interface TokenInfo {
  tokenId: number;
  text: string;
  frequency: number;
  isSpecial: boolean;
  is_special: boolean;
  startPosition: number | null;
  endPosition: number | null;
  start_position: number | null;
  end_position: number | null;
}

/**
 * Summary statistics about the vocabulary.
 */
export interface VocabularySummary {
  /** Total number of tokens in the vocabulary */
  vocabSize: number;
  /** Special token mappings */
  specialTokens: SpecialTokens;
  /** Most frequent tokens in the vocabulary */
  mostCommon: Array<{
    id: number;
    text: string;
    frequency: number;
  }>;
}

/**
 * Special token IDs used by the tokenizer.
 */
export interface SpecialTokens {
  /** Padding token ID */
  pad: number;
  /** Unknown token ID (for out-of-vocabulary tokens) */
  unk: number;
  /** Start of sequence token ID */
  sos: number;
  /** End of sequence token ID */
  eos: number;
}

/**
 * Complete vocabulary data structure.
 */
export interface VocabularyData {
  /** Total size of the vocabulary */
  size: number;
  /** All tokens in the vocabulary */
  tokens: TokenData[];
  /** Type of tokenizer used */
  tokenizerType: TokenType;
  /** Special token mappings */
  specialTokens: SpecialTokens;
}

/**
 * Result of tokenizing a text.
 */
export interface TokenizationResult {
  /** List of token IDs */
  tokenIds: number[];
  /** List of token strings (in the same order as token IDs) */
  tokens: string[];
  /** Detailed information about each token */
  tokenDetails: TokenInfo[];
  /** Summary of the vocabulary used */
  vocabularySummary: VocabularySummary;
  /** Type of tokenizer that was used */
  tokenizerType: TokenType;
}

/**
 * Request to tokenize text.
 */
export interface TokenizeRequest {
  /** Text to tokenize */
  text: string;
  /** Type of tokenizer to use */
  tokenizerType: TokenType;
  /** Maximum vocabulary size */
  vocabSize?: number;
  /** Number of BPE merge operations (only for BPE tokenizer) */
  numMerges?: number;
  /** Optional training texts (if not provided, uses default corpus) */
  trainingTexts?: string[];
}

/**
 * Response from the tokenize endpoint.
 */
export interface TokenizeResponse {
  /** Whether the tokenization was successful */
  success: boolean;
  /** List of token IDs */
  tokenIds: number[];
  /** List of token strings */
  tokens: string[];
  /** Detailed token information */
  tokenDetails: TokenInfo[];
  /** Vocabulary summary */
  vocabularySummary: VocabularySummary;
  /** Type of tokenizer used */
  tokenizerType: TokenType;
  /** Error message if failed */
  error?: string;
}

/**
 * Request to get vocabulary.
 */
export interface VocabularyRequest {
  /** Type of tokenizer */
  tokenizerType: TokenType;
  /** Maximum vocabulary size */
  vocabSize?: number;
  /** Optional training texts */
  trainingTexts?: string[];
}

/**
 * Response from the vocabulary endpoint.
 */
export interface VocabularyResponse {
  /** Whether the request was successful */
  success: boolean;
  /** Total vocabulary size */
  vocabSize: number;
  /** All tokens in the vocabulary */
  tokens: Array<{
    id: number;
    text: string;
    frequency: number;
    isSpecial: boolean;
  }>;
  /** Special token mappings */
  specialTokens: SpecialTokens;
  /** Type of tokenizer */
  tokenizerType: TokenType;
  /** Error message if failed */
  error?: string;
}

/**
 * Request to decode token IDs.
 */
export interface DecodeRequest {
  /** Token IDs to decode */
  tokenIds: number[];
  /** Type of tokenizer */
  tokenizerType: TokenType;
  /** Maximum vocabulary size */
  vocabSize?: number;
  /** Optional training texts */
  trainingTexts?: string[];
}

/**
 * Response from the decode endpoint.
 */
export interface DecodeResponse {
  /** Whether decoding was successful */
  success: boolean;
  /** Decoded text */
  text: string;
  /** Type of tokenizer used */
  tokenizer_type: TokenType;
  /** Error message if failed */
  error?: string;
}

/**
 * Request to compare tokenizers.
 */
export interface CompareRequest {
  /** Text to tokenize */
  text: string;
  /** Maximum vocabulary size for all tokenizers */
  vocabSize?: number;
  /** Number of BPE merge operations */
  numMerges?: number;
  /** Optional training texts */
  trainingTexts?: string[];
}

/**
 * Result from comparing different tokenizers.
 */
export interface ComparisonResult {
  /** Type of tokenizer */
  tokenizerType: TokenType;
  /** Number of tokens produced */
  numTokens: number;
  /** Token strings */
  tokens: string[];
  /** Token IDs */
  tokenIds: number[];
  /** Number of out-of-vocabulary tokens */
  oovCount: number;
}

/**
 * Response from the compare endpoint.
 */
export interface CompareResponse {
  /** Whether comparison was successful */
  success: boolean;
  /** Comparison results for each tokenizer */
  comparisons: ComparisonResult[];
  /** Error message if failed */
  error?: string;
}

/**
 * Tokenizer state in the global store.
 */
export interface TokenizerState {
  /** Currently selected tokenizer type */
  tokenizerType: TokenType;
  /** Maximum vocabulary size */
  vocabSize: number;
  /** Number of BPE merges (for BPE tokenizer) */
  numMerges: number;
  /** Current vocabulary data */
  vocabularyData: VocabularyData | null;
  /** Latest tokenization result */
  tokenizationResult: TokenizationResult | null;
  /** Whether a tokenization is in progress */
  isLoading: boolean;
  /** Error message if tokenization failed */
  error: string | null;
}

/**
 * Tokenizer configuration options.
 */
export interface TokenizerConfig {
  /** Type of tokenizer */
  type: TokenType;
  /** Maximum vocabulary size */
  vocabSize: number;
  /** Number of BPE merges (only for BPE) */
  numMerges?: number;
  /** Whether to use pretrained vocabulary */
  usePretrained?: boolean;
}

/**
 * Token embedding data for visualization.
 */
export interface TokenEmbeddingData {
  /** Token ID */
  tokenId: number;
  /** Token text */
  text: string;
  /** Embedding vector */
  embedding: number[];
  /** 2D projection coordinates (for visualization) */
  projection?: {
    x: number;
    y: number;
  };
}

/**
 * Frequency distribution data for visualization.
 */
export interface FrequencyDistributionData {
  /** Token texts */
  labels: string[];
  /** Frequency values */
  frequencies: number[];
  /** Token IDs */
  tokenIds: number[];
}

/**
 * Search/filter options for vocabulary browser.
 */
export interface VocabularySearchOptions {
  /** Search query (matches token text) */
  query: string;
  /** Minimum frequency */
  minFrequency?: number;
  /** Maximum frequency */
  maxFrequency?: number;
  /** Whether to include special tokens */
  includeSpecialTokens: boolean;
  /** Sort order */
  sortBy: 'id' | 'frequency' | 'text';
  /** Sort direction */
  sortOrder: 'asc' | 'desc';
  /** Maximum number of results */
  limit?: number;
  /** Offset for pagination */
  offset?: number;
}

/**
 * Filtered vocabulary results.
 */
export interface FilteredVocabularyResult {
  /** Total number of matching tokens */
  total: number;
  /** Filtered tokens */
  tokens: TokenData[];
  /** Whether there are more results */
  hasMore: boolean;
}

/**
 * Tokenizer comparison metrics.
 */
export interface ComparisonMetrics {
  /** Average token sequence length */
  avgSequenceLength: number;
  /** Average vocabulary size used */
  avgVocabSize: number;
  /** Average OOV rate */
  avgOovRate: number;
  /** Compression ratio (tokens per character) */
  compressionRatio: number;
}

/**
 * Tokenizer health/availability information.
 */
export interface TokenizerHealthInfo {
  /** Whether the tokenizer is available */
  available: boolean;
  /** Supported tokenizer types */
  supportedTypes: TokenType[];
  /** Default tokenizer type */
  defaultType: TokenType;
  /** Maximum allowed vocabulary size */
  maxVocabSize: number;
  /** Minimum allowed vocabulary size */
  minVocabSize: number;
}

/**
 * Training corpus statistics.
 */
export interface CorpusStatistics {
  /** Total number of texts */
  textCount: number;
  /** Total number of characters */
  charCount: number;
  /** Estimated number of words */
  wordCount: number;
  /** Most frequent characters */
  topChars: Array<{ char: string; count: number }>;
  /** Most frequent words */
  topWords: Array<{ word: string; count: number }>;
}

/**
 * Event types for tokenizer-related events.
 */
export type TokenizerEventType =
  | 'tokenizer-type-changed'
  | 'vocabulary-loaded'
  | 'tokenization-complete'
  | 'tokenization-error'
  | 'tokenizer-trained';

/**
 * Tokenizer event payload.
 */
export interface TokenizerEvent {
  type: TokenizerEventType;
  payload: {
    tokenizerType?: TokenType;
    timestamp: number;
    data?: unknown;
  };
}

/**
 * Token color mapping for visualization.
 * Maps token types or properties to colors.
 */
export interface TokenColorScheme {
  /** Color for special tokens */
  special: string;
  /** Color for regular tokens */
  regular: string;
  /** Color for unknown/out-of-vocabulary tokens */
  unknown: string;
  /** Color for highlighted tokens */
  highlighted: string;
  /** Background color for tokens */
  background: string;
}

/**
 * Token visualization options.
 */
export interface TokenVisualizationOptions {
  /** Whether to show token IDs */
  showIds: boolean;
  /** Whether to show token frequencies */
  showFrequencies: boolean;
  /** Whether to show special tokens */
  showSpecialTokens: boolean;
  /** Whether to show position information */
  showPositions: boolean;
  /** Whether to show embeddings */
  showEmbeddings: boolean;
  /** Color scheme to use */
  colorScheme: TokenColorScheme;
  /** Maximum number of tokens to display */
  maxTokens?: number;
}